dftimewolf/cli/curses_display_manager.py

```diff
@@ -0,0 +1,290 @@
+# -*- coding: utf-8 -*-
+"""Curses output management class."""
+
+from enum import Enum
+import threading
+import traceback
+from typing import Any, Dict, List, Optional, Union
+
+import curses
+
+class Status(Enum):
+  """Enum class for module states.
+
+  The order here is important, as it's the order modules will be displayed."""
+  COMPLETED = 'Completed'
+  SETTINGUP = 'Setting Up'
+  ERROR = 'Error'
+  RUNNING = 'Running'
+  PREPROCESSING = 'Preprocessing'
+  PROCESSING = 'Processing'
+  POSTPROCESSING = 'Postprocessing'
+  PENDING = 'Pending'
+  CANCELLED = 'Cancelled'
+
+
+class Module:
+  """An object used by the CursesDisplayManager used to represent a DFTW module.
+  """
+  def __init__(self,
+               name: str,
+               dependencies: List[str],
+               runtime_name: Optional[str] = None):
+    """Initialise the Module object."""
+    self.name = name
+    self.runtime_name = runtime_name if runtime_name else name
+    self.status: Status = Status.PENDING
+    self._dependencies: List[str] = dependencies
+    self._error_message: str = ''
+    self._threads: Dict[str, Dict[str, Any]] = {}
+    self._threads_containers_max: int = 0
+    self._threads_containers_completed: int = 0
+
+  def Stringify(self) -> List[str]:
+    """Returns an CursesDisplayManager friendly string describing the module."""
+    module_line = f'     {self.runtime_name}: {self.status.value}'
+    thread_lines = []
+
+    if self.status == Status.PENDING and len(self._dependencies) != 0:
+      module_line += f' ({", ".join(self._dependencies)})'
+    elif self.status == Status.ERROR:
+      module_line += f': {self._error_message}'
+    elif self.status in [Status.RUNNING, Status.PROCESSING] and self._threads:
+      module_line += (f' - {self._threads_containers_completed} of '
+          f'{self._threads_containers_max} containers completed')
+      for n, t in self._threads.items():
+        thread_lines.append(
+            f'       {n}: {t["status"].value} ({t["container"]})')
+
+    return [module_line] + thread_lines
+
+  def SetStatus(self, status: Status) -> None:
+    """Set the status of the module."""
+    if self.status not in [Status.ERROR, Status.COMPLETED, Status.CANCELLED]:
+      self.status = status
+
+  def SetThreadState(self, thread: str, status: Status, container: str) -> None:
```

Comment:
Could you have a bit more detailed (e.g. adding an `Args:` sections) docstrings here and throughout the PR? I'm struggling to understand what e.g. the container str is used for, or should be.

Reply:
Added.